packages/node/src/indexer/apiPromise.connection.spec.ts

```diff
@@ -0,0 +1,68 @@
+// Copyright 2020-2022 OnFinality Limited authors & contributors
+// SPDX-License-Identifier: Apache-2.0
+
+import { ApiPromise } from '@polkadot/api';
+import { WsProvider, HttpProvider } from '@polkadot/rpc-provider';
+import { ApiPromiseConnection } from './apiPromise.connection';
+
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const { version: packageVersion } = require('../../package.json');
+const RETRY_DELAY = 2_500;
+
+const TEST_BLOCKHASH =
+  '0x3b712c10ba98c5421a468d1411c94479381be307b7ecb5b2602cf0d395eb4292';
+
+describe('ApiPromiseConnection', () => {
+  let wsProvider;
+  let httpProvider;
+
+  beforeEach(async () => {
+    wsProvider = new WsProvider('wss://kusama.api.onfinality.io/public-ws');
+    httpProvider = new HttpProvider('https://kusama.api.onfinality.io/public');
+
+    const api = await ApiPromise.create({ provider: wsProvider });
+
+    jest.spyOn(wsProvider, 'send');
+    jest.spyOn(httpProvider, 'send');
+  });
+
+  it('should not make duplicate requests for state_getRuntimeVersion on wsProvider', async () => {
+    const cachedProvider =
+      ApiPromiseConnection.createCachedProvider(wsProvider);
+
+    await cachedProvider.send('state_getRuntimeVersion', [TEST_BLOCKHASH]);
+    await cachedProvider.send('state_getRuntimeVersion', [TEST_BLOCKHASH]);
+
+    expect(wsProvider.send).toHaveBeenCalledTimes(1);
```

Comment:
All of these tests would fail if we run multiple requests at once. I know the usage is not like this currently but with a simple change we can also support this use case

e.g failing test
```
    await Promise.all([
          cachedProvider.send('state_getRuntimeVersion', [TEST_BLOCKHASH]),
          cachedProvider.send('state_getRuntimeVersion', [TEST_BLOCKHASH]),
    ]);

    expect(wsProvider.send).toHaveBeenCalledTimes(1);
```

To solve this we should cache the promise instead of the result